tests/integration/sync/sync_integ_base.py

```diff
@@ -1,19 +1,151 @@
+        self.lambda_client = boto3.client("lambda")
+        self.api_client = boto3.client("apigateway")
+        self.sfn_client = boto3.client("stepfunctions")
+        self.sns_arn = os.environ.get("AWS_SNS")
+        self.stacks = []
+        time.sleep(CFN_SLEEP)
+        self.s3_prefix = uuid.uuid4().hex
         super().setUp()
 
     def tearDown(self):
+        shutil.rmtree(os.path.join(os.getcwd(), ".aws-sam", "build"), ignore_errors=True)
+        for stack in self.stacks:
+            # because of the termination protection, do not delete aws-sam-cli-managed-default stack
+            stack_name = stack["name"]
+            if stack_name != SAM_CLI_STACK_NAME:
+                region = stack.get("region")
+                cfn_client = (
+                    self.cfn_client if not region else boto3.client("cloudformation", config=Config(region_name=region))
+                )
+                ecr_client = self.ecr_client if not region else boto3.client("ecr", config=Config(region_name=region))
+                self._delete_companion_stack(cfn_client, ecr_client, self._stack_name_to_companion_stack(stack_name))
+                cfn_client.delete_stack(StackName=stack_name)
         super().tearDown()
 
+    def _get_stacks(self, stack_name):
+        try:
+            physical_ids = {}
+            response = self.cfn_client.describe_stack_resources(StackName=stack_name).get("StackResources", {})
+            for resource in response:
+                resource_type = resource.get("ResourceType")
+                if resource_type == "AWS::CloudFormation::Stack":
+                    nested_stack_physical_id = resource.get("PhysicalResourceId")
+                    nested_stack_name = nested_stack_physical_id.split("/")[1]
+                    nested_stack_physical_ids = self._get_stacks(nested_stack_name)
+                    for nested_resource_type, nested_physical_ids in nested_stack_physical_ids.items():
+                        if nested_resource_type not in physical_ids:
+                            physical_ids[nested_resource_type] = []
+                        physical_ids[nested_resource_type] += nested_physical_ids
+                    continue
+                if resource_type not in physical_ids:
+                    physical_ids[resource.get("ResourceType")] = []
+                physical_ids[resource_type].append(resource.get("PhysicalResourceId"))
+            return physical_ids
+        except ClientError:
+            pass
+        return None
+
+    def _get_lambda_response(self, lambda_function):
+        try:
+            lambda_response = self.lambda_client.invoke(FunctionName=lambda_function, InvocationType="RequestResponse")
+            payload = json.loads(lambda_response.get("Payload").read().decode("utf-8"))
+            return payload.get("body")
+        except ClientError as ce:
+            LOG.error(ce)
+        except Exception as e:
+            LOG.error(e)
```

Comment:
Changing all error to directly raise, the only concern that I have is that it would increase the frequency which infra integration tests fail. With the while loop attempt if any one of the attempts gets an client error the whole test case simply fails. For now let's just raise any error and observe the stability of these tests.